Fan-in Benchmark

ReubenMathew opened the pull request:
Add new JetStream fan-in benchmark

~~Some cleanup for the existing core fan-in benchmark.~~

@mprimi do you have any specific suggestions regarding making this look cleaner?

ReubenMathew committed:
Add JetStream multi-producer benchmark

Benchmark measures the performance of multiple connections publishing
bytes concurrently to a stream.

Signed-off-by: reubenninan <reuben@nats.io>

neilalexander committed:
Reduce reallocations in `metaSnapshot`

Signed-off-by: Neil Twigg <neil@nats.io>

neilalexander committed:
Reduce reflection on marshal, reduce string building on unmarshal

Signed-off-by: Neil Twigg <neil@nats.io>

neilalexander committed:
Predefine byte copies of frequently-used consts

Signed-off-by: Neil Twigg <neil@nats.io>

derekcollison committed:
Optimise metalayer snapshots (#4925)

This PR optimises the marshalling of metalayer snapshots with a few
techniques:

* Simplifies the custom marshalers for a number of types that appear
often in the snapshotted config, as these were using reflection
unnecessarily
* Simplifies the custom unmarshallers for the same types, as these were
string-building more than necessary
* Attempts to reduce the number of reallocations that might happen in
`metaSnapshot()` by preallocating slices

Signed-off-by: Neil Twigg <neil@nats.io>

dependabot[bot] committed:
Bump golang.org/x/sys from 0.15.0 to 0.16.0

Bumps [golang.org/x/sys](https://github.com/golang/sys) from 0.15.0 to 0.16.0.
- [Commits](https://github.com/golang/sys/compare/v0.15.0...v0.16.0)

---
updated-dependencies:
- dependency-name: golang.org/x/sys
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

neilalexander committed:
Don't allow stream or consumer snapshots before recovery completed

Signed-off-by: Neil Twigg <neil@nats.io>

derekcollison committed:
Bump golang.org/x/sys from 0.15.0 to 0.16.0 (#4926)

Bumps [golang.org/x/sys](https://github.com/golang/sys) from 0.15.0 to
0.16.0.
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/golang/sys/commit/0829ab15b6946f47c40012db2e0c04772730317d"><code>0829ab1</code></a>
windows: add SetFileValidData</li>
<li><a
href="https://github.com/golang/sys/commit/32cdffc66120e843f9768b765d98b513ce77d6a7"><code>32cdffc</code></a>
unix: don't redefine constants already defined in glibc headers</li>
<li><a
href="https://github.com/golang/sys/commit/f0c7190c5ee2c4ec638b0e505866bb6456fc83c4"><code>f0c7190</code></a>
unix: remove extra trailing newlines in zsyscall_openbsd_*.go</li>
<li><a
href="https://github.com/golang/sys/commit/5ff87d7bfdb95967279073074ad0ef6f473f3e5c"><code>5ff87d7</code></a>
unix: add Netfilter and NFTables constants</li>
<li>See full diff in <a
href="https://github.com/golang/sys/compare/v0.15.0...v0.16.0">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=golang.org/x/sys&package-manager=go_modules&previous-version=0.15.0&new-version=0.16.0)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore this major version` will close this PR and stop
Dependabot creating any more for this major version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this minor version` will close this PR and stop
Dependabot creating any more for this minor version (unless you reopen
the PR or upgrade to it yourself)
- `@dependabot ignore this dependency` will close this PR and stop
Dependabot creating any more for this dependency (unless you reopen the
PR or upgrade to it yourself)


</details>

derekcollison committed:
Don't allow stream or consumer snapshots before recovery completed (#4927)

Since `doSnapshot` in both cases can be called from a maintenance timer
or on the quit signal, this may cause us to snapshot before recovery has
completed.

Signed-off-by: Neil Twigg <neil@nats.io>

ReubenMathew commented:
Example run of new JS fan-in:
```
❯ go test --bench 'BenchmarkJetStreamFanIn/.*' -run=^$ --count 1
goos: darwin
goarch: arm64
pkg: github.com/nats-io/nats-server/v2/server
BenchmarkJetStreamFanIn/msgSz=100b/pubs=3-8  	  20132	    59339 ns/o   1.69 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=100b/pubs=5-8  	  27537	    43500 ns/o   2.30 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=100b/pubs=10-8 	  35652	    32080 ns/o   3.12 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=1024b/pubs=3-8 	  17856	    67441 ns/o  15.18 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=1024b/pubs=5-8 	  23035	    50916 ns/o  20.11 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=1024b/pubs=10-8         	  28737	    41754 ns/op	 24.52 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=10240b/pubs=3-8         	  11018	    99051 ns/op	103.38 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=10240b/pubs=5-8         	  14900	    78371 ns/op	130.66 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=10240b/pubs=10-8        	  15297	    71652 ns/op	142.91 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=524288b/pubs=3-8        	    957	  1183739 ns/op	442.91 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=524288b/pubs=5-8        	    997	  1157081 ns/op	453.11 MB/s	        0 %error
BenchmarkJetStreamFanIn/msgSz=524288b/pubs=10-8       	    960	  1864563 ns/op	281.19 MB/s	        0 %error
PASS
ok  	github.com/nats-io/nats-server/v2/server	77.536s
```

mprimi commented on server/jetstream_benchmark_test.go:
```diff
@@ -1295,6 +1295,187 @@ func BenchmarkJetStreamObjStore(b *testing.B) {
+		publishersReadyWg.Add(numPubs + 1)
+
+		// create N publishers
+		publishers := func() []BenchPublisher {
```
Here you are just creating 10 connections and 10 channels. 
You don't need a function and you don't need a waitgroup.. 
Just create 10 publisher in a for loop?

mprimi commented on server/jetstream_benchmark_test.go:
```diff
@@ -1295,6 +1295,187 @@ func BenchmarkJetStreamObjStore(b *testing.B) {
+				for {
+					select {
+					case <-publishers[pubId].quitCh:
+						publishers[pubId].conn.Close()
```
Safer to close using `defer`, either in the publisher goroutine, or in the init block above when the connnections are created.

mprimi commented on server/jetstream_benchmark_test.go:
```diff
@@ -1295,6 +1295,187 @@ func BenchmarkJetStreamObjStore(b *testing.B) {
+					pubAck, err := publishers[pubId].js.Publish(subject, messageData)
+					if err != nil {
+						publishers[pubId].publishErrors += 1
+						fmt.Println(err)
```
Use log so it's printed only when tests are run with `-v`

mprimi commented on server/jetstream_benchmark_test.go:
```diff
@@ -1295,6 +1295,187 @@ func BenchmarkJetStreamObjStore(b *testing.B) {
+					}
+
+					// all messages have been published to stream
+					if int(pubAck.Sequence) == b.N {
```
Doesn't matter but as general rule it's probably safer to cast the smallest type to the larger, so cast `b.N`

mprimi replied:
If you change this check to: `if pubAck.Sequence > b.N  then return` you can simplify your code quite a bit, no need for a quit channel.

Slightly changes what you are measuring, i.e. "time for everyone to see a puback > b.N".

You could bring it back to "time for stream to reach b.N" with some additional mechanism triggered by just 1 (the first) producer.

mprimi commented on server/jetstream_benchmark_test.go:
```diff
@@ -1295,6 +1295,187 @@ func BenchmarkJetStreamObjStore(b *testing.B) {
+		for _, publisher := range publishers {
+			totalErrors += publisher.publishErrors
+		}
+		errorRate := 100 * float64(totalErrors) / float64(b.N)
```
The denominator here is not exactly right. Should be number of total calls to `Publish`

```
N 3
Publish calls 10
Errors 7

Current error rate: 230%
Actual error rate: 70%
```

mprimi commented on server/jetstream_benchmark_test.go:
```diff
@@ -1295,6 +1295,187 @@ func BenchmarkJetStreamObjStore(b *testing.B) {
+
+	fanIn := func(b *testing.B, numPubs int, messageSize int64) {
+
+		cl, ls, shutdown, nc, js := startJSClusterAndConnect(b, clusterSize)
```
I'd leave cluster creation (maybe stream too) down in the "main" loop, and just have benchmark code here in this benchmark function.

mprimi commented on server/core_benchmarks_test.go:
```diff
@@ -427,157 +426,161 @@ func BenchmarkCoreFanIn(b *testing.B) {
+						fmt.Sprintf("pubs=%d", numPubs),
+						func(b *testing.B) {
+							// run workload
+							fanInWorkload(b, numPubs, messageSize)
```
There's 2 changes in 1 commit here which make this hard to review:
 - Move to a benchmark function to simplify the "main"
 - Adjust benchmark logic

I would say, forget about 1 for now, focus on 2.

Later and/or in a separate commit, take care of 1 for a all tests in this file.

(also valid: take care of 1 for the whole file, verify nothing changes just style, then move on to make logic changes)

I have not reviewed this file in detail because of the above, but i suspect a number of comments i made on the JS bench below also apply here. Especially setup and wrap up logic.

ReubenMathew commented on server/jetstream_benchmark_test.go:
```diff
@@ -1295,6 +1295,187 @@ func BenchmarkJetStreamObjStore(b *testing.B) {
 	}
 }
 
+func BenchmarkJetStreamFanIn(b *testing.B) {
```
`JetStreamMultiPublish`?

derekcollison committed:
Optimize catchups for replicas and mirrors with large interior deletes.

Under scenarios where interior deletes were orders of magnitudes larger than the msgs themselves, the system would need to transfer all interior deletes as skip msgs.
This change adopts the ability to send DeleteRanges on streamSyncRequests, and for R1 mirrors to locally use the new store.SkipMsgs to optimize speed.

This CL also has a bug fix for memory store state with delete details, and several general improvements to the filestore impl.

Signed-off-by: Derek Collison <derek@nats.io>

derekcollison committed:
Fix for not reporting deleted msgs past last after a truncate

Signed-off-by: Derek Collison <derek@nats.io>

derekcollison committed:
[IMPROVED] Optimize catchups for replicas and mirrors with large interior deletes. (#4929)

Under scenarios where interior deletes were orders of magnitude larger
than the msgs themselves, the system would need to transfer all interior
deletes as skip msgs. This change adopts the ability to send
DeleteRanges on streamSyncRequests, and for R1 mirrors to locally use
the new store.SkipMsgs to optimize speed.

This CL also has a bug fix for memory store state with delete details,
and several general improvements to the filestore impl.

Signed-off-by: Derek Collison <derek@nats.io>

levb committed:
[IMPROVED] MQTT: concurrently fetch retained messages for a new SUB (#4835)

Speed up sending out retained messages to a sub by loading them from
JetStream concurrently.

- added `jsa.newRequestExMulti` that fans out multiple requests (on the
same subject, at least for now) and gathers all responses
- further changed the processing of retained messages in
`as.processSubs`: now we
- a) find all messages that we need and that are not in `rmsCache` first
  - (b) load them to `rms` with one call
- (c) serialize to subs/filters and fail the filter if it misses
messages from (a) in `rms`.

---------

Signed-off-by: Ivan Kozlovic <ivan@synadia.com>
Co-authored-by: Ivan Kozlovic <ivan@synadia.com>

derekcollison committed:
[IMPROVED] Avoid stream locks for certain operations that could block routes/gateways. (#4933)

Avoid needing the lock for mset subscriptions to avoid possibly blocking
routes/gateways.
Also increase timeout for consumer creation for sources since might take
a bit longer.

 Signed-off-by: Derek Collison <derek@nats.io>

---------

Signed-off-by: Derek Collison <derek@nats.io>

jnmoyne committed:
Finds the starting sequence number for all of the streams sources in one go rather than repeatedly
Consolidation to avoid code duplication between setStartingSequenceForSource and setStartingSequenceForSources

Signed-off-by: Jean-Noël Moyne <jnmoyne@gmail.com>

derekcollison committed:
When reloads contain mapping updates for global account with leadnode connections correctly update their smaps.

Signed-off-by: Derek Collison <derek@nats.io>

derekcollison committed:
[FIXED] Reloads that contain mapping updates for global account with leafnodes. (#4937)

We should update the leafnode connection's smaps correctly.

Resolves: #4934 

Signed-off-by: Derek Collison <derek@nats.io>

derekcollison committed:
Allow no_auth_user to be an nkey.

Signed-off-by: Derek Collison <derek@nats.io>

derekcollison committed:
Improve finding the last sourced message for each source of a stream (#4935)

Improves how the sequence number of the last sourced message is found
for all the stream's sources (in a single pass rather that individual
passes for each source).

Signed-off-by: Jean-Noël Moyne <jnmoyne@gmail.com>

derekcollison committed:
[IMPROVED] Allow no_auth_user to be an nkey. (#4938)

Partially resolves #3196 

Signed-off-by: Derek Collison <derek@nats.io>

derekcollison committed:
Allow leafnode authorization to support nkeys.

Signed-off-by: Derek Collison <derek@nats.io>

derekcollison committed:
[FIXED] Allow leafnode authorization to support nkeys. (#4940)

Resolves: #3196 

Signed-off-by: Derek Collison <derek@nats.io>

ripienaar committed:
Formalize claims update response

The previous response was attempting to look like a typical
Api response but in the process invented its own way of
representing error and more making it really hard for
users to figure out how this is supposed to look.

Added a struct that retains the non standard behavior
while working on CLI tooling

ReubenMathew commented:
@mprimi sample run
```
goos: darwin
goarch: arm64
pkg: github.com/nats-io/nats-server/v2/server
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=100b/pubs=3-8  	   52582	     21408 ns/op	   4.67 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=100b/pubs=5-8  	   73107	     15189 ns/op	   6.58 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=100b/pubs=10-8 	   93260	     12587 ns/op	   7.94 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=1024b/pubs=3-8 	   49594	     25997 ns/op	  39.39 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=1024b/pubs=5-8 	   57994	     28888 ns/op	  35.45 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=1024b/pubs=10-8         	   55286	     22709 ns/op	  45.09 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=10240b/pubs=3-8         	   30326	     37903 ns/op	 270.17 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=10240b/pubs=5-8         	   36274	     39944 ns/op	 256.36 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=10240b/pubs=10-8        	   36393	     33092 ns/op	 309.44 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=524288b/pubs=3-8        	    2065	    557340 ns/op	 940.70 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=524288b/pubs=5-8        	    3434	    510639 ns/op	1026.73 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=1,R=1/msgSz=524288b/pubs=10-8       	    3312	    477439 ns/op	1098.13 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=100b/pubs=3-8           	   20218	     59331 ns/op	   1.69 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=100b/pubs=5-8           	   26414	     44088 ns/op	   2.27 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=100b/pubs=10-8          	   35582	     31495 ns/op	   3.18 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=1024b/pubs=3-8          	   15946	     67998 ns/op	  15.06 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=1024b/pubs=5-8          	   22732	     50606 ns/op	  20.23 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=1024b/pubs=10-8         	   30550	     38708 ns/op	  26.45 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=10240b/pubs=3-8         	   11434	    103746 ns/op	  98.70 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=10240b/pubs=5-8         	   14192	     85612 ns/op	 119.61 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=10240b/pubs=10-8        	   17300	     69847 ns/op	 146.61 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=524288b/pubs=3-8        	     836	   1242354 ns/op	 422.01 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=524288b/pubs=5-8        	    1002	   1196019 ns/op	 438.36 MB/s	         0 %error
BenchmarkJetStreamMultiProducer/N=3,R=3/msgSz=524288b/pubs=10-8       	    1132	   1210450 ns/op	 433.13 MB/s	         0 %error
PASS
ok  	github.com/nats-io/nats-server/v2/server	94.314s
```

wallyqs committed:
Always restart OCSP monitors on reload

Signed-off-by: Waldemar Quevedo <wally@nats.io>

derekcollison committed:
Formalize claims update response (#4942)

The previous response was attempting to look like a typical Api response
but in the process invented its own way of representing error and more
making it really hard for users to figure out how this is supposed to
look.

Added a struct that retains the non standard behavior while working on
CLI tooling

Signed-off-by: R.I.Pienaar <rip@devco.net>

derekcollison committed:
Always restart OCSP monitors on reload (#4943)

Instead of attempting to detect whether the TLS configuration has
changed
on each tls connection type, this changes to always restart the OCSP
monitors
on reload in order to avoid `peer missing OCSP Staple` errors.

wallyqs committed:
Bump Go version

Signed-off-by: Waldemar Quevedo <wally@nats.io>

jnmoyne committed:
Increase timing of the checks at the end of the test

Signed-off-by: Jean-Noël Moyne <jnmoyne@gmail.com>

derekcollison committed:
Update crypto dependency

Signed-off-by: Derek Collison <derek@nats.io>

derekcollison committed:
Bump Go version to 1.21.6 (#4945)

Test versions are now:
 - 1.21.6
 - 1.20.13

derekcollison committed:
Update crypto dependency (#4946)

Signed-off-by: Derek Collison <derek@nats.io>

derekcollison committed:
Fix flapping in TestJetStreamMirrorBasics (#4947)

It can take up to a couple of seconds for new source consumers to get
established and all the message getting sourced in the test, the final
tests had a timeout of 2 seconds that could sometimes be reached in
Travis. Relaxing that timeout to something more generous fixes the
flapping

ReubenMathew committed:
Fix: core fan-in benchmark improvements

Benchmark was previously not properly waiting for publisher
sub-routines to be created before starting fan-in workload.

Publishers now generate a new random payload for each publish iteration.

Style and formatting fixes to increase readability.

Signed-off-by: reubenninan <reuben@nats.io>

jnmoyne committed:
Fix line with wrong variable add extra guard and optimization

Signed-off-by: Jean-Noël Moyne <jnmoyne@gmail.com>

derekcollison committed:
[FIXED] setStartingSequenceForSources could panic in some specific scenarios (#4950)

Fixed setStartingSequenceForSources line using the wrong variable name

mprimi commented:
Feedback on Core: 
 - Logic around `publishersReadyWg` is not right
 - Like the JS bench, this stops at `numPubs * b.N`, so it will appear to be slower with more publishers
 - Publish some sample result on next iteration of this PR
 - Title of commit is misleading, it's clearly doing more than style and formatting

For future, better to have style/format in one commit, should behave same as before.
And then logic change on the next commit, easier to review.
